src/QueryMiddleware.php

```diff
@@ -82,35 +84,83 @@ public function __construct(
 
     public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
     {
-        $queryName = $request->getAttribute(self::NAME_ATTRIBUTE);
+        $messages = $this->parseRequestMessages($request);
 
-        if (null === $queryName) {
-            throw new RuntimeException(
-                sprintf('Query name attribute ("%s") was not found in request.', self::NAME_ATTRIBUTE),
-                StatusCodeInterface::STATUS_BAD_REQUEST
+        $responses = [];
+
+        foreach ($messages as $message) {
+            $message['metadata'] = $this->metadataGatherer->getFromRequest($request);
+
+            $query = $this->queryFactory->createMessageFromArray(
+                $message[self::NAME_ATTRIBUTE],
+                $message
             );
-        }
-        $payload = $request->getQueryParams();
 
-        if ($request->getMethod() === RequestMethodInterface::METHOD_POST) {
-            $payload['data'] = $request->getParsedBody();
+            try {
+                $responses[] = $this->queryBus->dispatch($query);
+            } catch (\Throwable $e) {
+                throw new RuntimeException(
+                    sprintf('An error occurred during dispatching of query "%s"', $message[self::NAME_ATTRIBUTE]),
+                    StatusCodeInterface::STATUS_INTERNAL_SERVER_ERROR,
+                    $e
+                );
+            }
         }
 
         try {
-            $query = $this->queryFactory->createMessageFromArray($queryName, [
-                'payload' => $payload,
-                'metadata' => $this->metadataGatherer->getFromRequest($request),
-            ]);
-
-            return $this->responseStrategy->fromPromise(
-                $this->queryBus->dispatch($query)
-            );
+            return $this->responseStrategy->fromPromise(all($responses));
         } catch (\Throwable $e) {
             throw new RuntimeException(
-                sprintf('An error occurred during dispatching of query "%s"', $queryName),
+                'An error occurred dispatching queries',
                 StatusCodeInterface::STATUS_INTERNAL_SERVER_ERROR,
                 $e
             );
         }
     }
+
+    private function parseRequestMessages(ServerRequestInterface $request): array
+    {
+        if ($request->getMethod() === RequestMethodInterface::METHOD_POST) {
```

Comment:
Instead of distinct between request methods, I would prefer to use other request attribute names.
In that case it would be possible to execute only one query by using a post body.

Reply:
There is nothing really stopping you from using a `POST` request to send only a single query with this pattern. I left the `GET` functionality in to support something closer to the original functionality and it does give a nice way to send just a single request, but I think I would suggest moving to just using the `$request->parsedBody()` if we wanted to simplify this code.

Reply:
I think it was a bad idea that I added `POST` for queries. Can't we just remove it? /cc @codeliner

Reply:
I've been playing around with PSR-7 request parsers. Specifically, I have been using Zend Diactoros. It looks like passing query parameter with a JSON input. It handles the query parameter parsing very well, so you can either pass in nested arrays or a single JSON object on the query, making it possible to not use `POST` but still handle multiple queries.

Nested:

`GET /events?queries[one][message_name]=query:find_user&queries[one][filter][]=John`

JSON

`GET /events?queries={"one":{"message_name":"query:find_user","filter":["John"]}`

Reply:
@prolic you need post requests, since some javascript packages always use POST for requests with body. Even if this is not part of what the RFC says, it was common many years to do it so.
GET for query params, POST for requests with body.

We should simply ignore the request method in this middleware and let the router decide to match. Also we need something to fetch the arguments from the request. E.g. I would to use "username" in payload for a route like `/users/{username}`

Reply:
from RFC 2616:

> if the request method
   does not include defined semantics for an entity-body, then the
   message-body SHOULD be ignored when handling the request.

https://tools.ietf.org/html/rfc2616#section-4.3

Let's see what defined for GET:

> The GET method means retrieve whatever information (in the form of an
   entity) is identified by the Request-URI.

https://tools.ietf.org/html/rfc2616#section-4.3

So the request-body is not part of the identification of the resource in a GET request, only the request URI.

So not only would I opt for removing the POST method here (because we query for something, so it's a GET request, as we are not sending data), but I would also opt for removing putting data in the request body

Reply:
GraphQL allows both GET and POST and nobody complains (me included)

First, we are not designing a REST-ish API here and RPC using POST requests does not feel wrong for me. It makes the client side simpler, too.

GET requests for queries make a lot of sense, bc only GET requests are cached normally. 
Disadvantage is that structured queries are harder to put into query params and the length is limited! So if we want to support multiple queries with arbitrary payload this might be a really issue!

Reply:
For reference, GraphQL implements using both the `GET` and `POST` verbs by actually passing in a single `query` property, either as a query parameter or a JSON body argument. The property is a JSON object defining the query being sent. It would most closely match my suggestion of 

`GET /events?queries={"one":{"message_name":"query:find_user","filter":["John"]}`

http://graphql.org/learn/serving-over-http/#get-request

Reply:
So are we fine with the expectation of a `POST` body consisting of the messages (plus identifiers) inside of `getParsedBody` as well as `GET` using `getQueryParams('queries')` and having that be the JSON body of the message? 

If so, I will make the changes and get the pushed to this PR either tonight or tomorrow.